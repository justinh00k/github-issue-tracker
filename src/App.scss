$big-font-size: 3vw;
$small-font-size: 1.25vw;
$border-radius: 0;

.App {
  text-align: center;
}

.App-logo {
  height: 30vw;
  pointer-events: none;
}

.loginInput {
  margin: 0 0 0 1.5vw;
  padding: 1.5vw;
  font-size: $big-font-size;
  border-radius: $border-radius;
  outline: 0;
}

.loginButton {

  cursor: pointer;
  height:  6vw;
  position: relative;
  top: .75vw;

}

.avatar {
  height: 4vw;
  max-width: 4vw;
}




@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-float infinite 2s ease-out;
  }
}

.loggedInWrapper {
  
  width: calc(100% - 10vw); 
    margin: 2.5vw 5vw;
}

.loginWrapper {
  
  margin-top: 2vw;


}

.subtitle {
font-size: $small-font-size
  
}
h2 
{
  margin-top: 0;
  font-size: $big-font-size;
  text-align: left;
}

.App-header {
  min-height: 25vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: $big-font-size;
}

.App-link {
  color: rgb(112, 76, 182);
}
.wrapper {
  display: flex; 
  flex-direction: row;
  border-radius: $border-radius;
  border: 1px solid #999; 
  font-size: $small-font-size;
  height: calc(94vh - 12vw);

}

.column {
  overflow: auto;
  border: 1px solid #999;

}

#issues tr:hover:not(:first-child) {
  background-color: #EEE;
}

#repos tr:hover {
  background-color: #EEE;
  cursor: pointer;
}

tr {
  height: 4vw;
  padding: 2vw;
}

#repos td {

  padding: 2vh;
}

table {
  width: 100%;
  border-spacing: 0;
}

th {
 
  cursor: pointer;
}

@keyframes App-logo-float {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.05);
  }
  100% {
    transform: scale(1);
  }
}


@media (max-width: 600px) {
  
  .wrapper {  
    flex-direction: column;
  }

  .column {
    border: 1px solid #999
  }



}