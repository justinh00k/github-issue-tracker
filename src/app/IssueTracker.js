import React, { useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCodeBranch,
  faSortUp,
  faSortDown,
} from "@fortawesome/free-solid-svg-icons";
import * as timeago from "timeago.js";
import { useSelector, useDispatch} from "react-redux";
import {getToken, getIssues, getIssuesAsync} from '../reducers/login'

const IssueTracker = ({ repos }) => {
  const token = useSelector(getToken);
  const issues = [...useSelector(getIssues)];
  const dispatch = useDispatch();
  const [sortOrder, setSortOrder] = useState(
    localStorage.getItem("sortOrder" + token) || ["created", false]
  );

  const sortIcon = (
    <FontAwesomeIcon
      icon={sortOrder[1] ? faSortDown : faSortUp}
      color="#222222"
      style={{ position: "absolute", marginLeft: "0.5vw" }}
      alt={`Sorted ${sortOrder[1] ? "Ascending" : "Descending"}`}
    />
  );

  const sortBy = (type) => {
    const newSort = [type, sortOrder[0] === type ? !sortOrder[1] : true];
    setSortOrder(newSort);
    localStorage.setItem("sortOrder" + token, newSort);
  };

  const getSortOrder = (a, b) => {
    switch (sortOrder[0]) {
      case "created":
        return sortOrder[1]
          ? new Date(a.created_at) - new Date(b.created_at)
          : new Date(b.created_at) - new Date(a.created_at);

      case "title":
        if (a.title < b.title) return sortOrder[1] ? -1 : 1;
        if (a.title > b.title) return sortOrder[1] ? 1 : -1;
        return 0;

      case "assignee":
        const assigneeA = a.assignee ? a.assignee.login : "";
        const assigneeB = b.assignee ? b.assignee.login : "";
        if (assigneeA < assigneeB) return sortOrder[1] ? -1 : 1;
        if (assigneeA > assigneeB) return sortOrder[1] ? 1 : -1;
        return 0;

      default:
      case "updated":
        return sortOrder[1]
          ? new Date(a.updated_at) - new Date(b.updated_at)
          : new Date(b.updated_at) - new Date(a.updated_at);
    }
  };

  return (
    <div className="loggedInWrapper">
      <h2>
        <FontAwesomeIcon
          icon={faCodeBranch}
          color="#CCC"
          size="1x"
          style={{ marginRight: "1vw" }}
          alt=""
        />
        GitHub Issue Tracker
        <span className="subtitle"> by justinh00k</span>
      </h2>

      <div className="wrapper">
        <div id="repos" className="column" style={{ flex: 1 }}>
          <table>
            <tbody>
              {repos.map((repo) => (
                <tr key={repo.id}>
                  <td
                    style={{
                      backgroundColor:
                        issues[0] && repo.url === issues[0].repository_url
                          ? "#CCC"
                          : "transparent",
                    }}
                    onClick={() => dispatch(getIssuesAsync(repo.url,token))}
                  >
                    {repo.full_name}
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>

        <div id="issues" className="column" style={{ flex: 3 }}>
          <table>
            <tbody>
              {issues.length > 0 && !!issues[0].id ? (
                <tr>
                  <th
                    style={{
                      textDecoration:
                        sortOrder[0] === "title" ? "underline" : "none",
                    }}
                    onClick={() => sortBy("title")}
                  >
                    Title{sortOrder[0] === "title" && sortIcon}
                  </th>
                  <th
                    style={{
                      width: "15%",
                      textDecoration:
                        sortOrder[0] === "created" ? "underline" : "none",
                    }}
                    onClick={() => sortBy("created")}
                  >
                    Created{sortOrder[0] === "created" && sortIcon}
                  </th>
                  <th
                    style={{
                      width: "15%",
                      textDecoration:
                        sortOrder[0] === "updated" ? "underline" : "none",
                    }}
                    onClick={() => sortBy("updated")}
                  >
                    Updated{sortOrder[0] === "updated" && sortIcon}
                  </th>
                  <th
                    style={{
                      width: "15%",
                      textDecoration:
                        sortOrder[0] === "assignee" ? "underline" : "none",
                    }}
                    onClick={() => sortBy("assignee")}
                  >
                    Assignee{sortOrder[0] === "assignee" && sortIcon}
                  </th>
                </tr>
              ) : (
                <tr>
                  <td colSpan={4}>
                    {issues[0] && !!issues[0].repository_url && (<b>No issues in this repo.</b>)}
                  </td>
                </tr>
              )}

              {issues[0] &&
                issues[0].id &&
                issues.sort(getSortOrder).map((issue) => (
                  <tr key={issue.id}>
                    <td>{issue.title}</td>
                    <td>{timeago.format(new Date(issue.created_at))}</td>
                    <td>{timeago.format(new Date(issue.updated_at))}</td>
                    <td>
                      {issue.assignee && (
                        <img
                          title={`Avatar of ${issue.assignee.login}`}
                          alt={`Avatar of ${issue.assignee.login}`}
                          className="avatar"
                          src={issue.assignee.avatar_url}
                        />
                      )}
                    </td>
                  </tr>
                ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  );
};

export default IssueTracker;