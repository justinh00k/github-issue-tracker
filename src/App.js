import React, { useState } from "react";

import "./App.scss";
import LoginScreen from './app/LoginScreen'
import IssueTracker from './app/IssueTracker'

function App() {
  const [repos, setRepos] = useState(false);
  const [token,setToken] = useState("")
  const [loginError,setLoginError] = useState(false)

  const login = async (loginToken) => {
    if (loginToken)
{
      setToken(loginToken)
      return fetch("https://api.github.com/user/repos", {
        method: "GET",
        headers: { Authorization: `token ${token}`},
      })
        .then((res) => res.json())
        .then((data) => {
          if (Array.isArray(data)){
          setRepos(data);
          setLoginError(false)
          }
          else setLoginError(true)
        })
        .catch(()=>{
          setLoginError(true)
        });
      }
      else
      setLoginError(true)
  };

  return (
    <div className="App">
      <header className="App-header">
        {repos 
        ? <IssueTracker token={token} repos={repos} /> 
        :   <LoginScreen loginError={loginError} setLoginError={setLoginError} login={login} />
        }
      </header>
    </div>
  );
}

export default App;
