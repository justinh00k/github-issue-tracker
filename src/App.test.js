import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import IssueTracker from './app/IssueTracker'

it('renders a snapshot', () => {
  const tree = renderer.create(<App/>).toJSON();
  expect(tree).toMatchSnapshot();
});

it('issue tracker loads even if user has no repos', () => {
  const tree = renderer.create(<IssueTracker user="" password="" repos={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
