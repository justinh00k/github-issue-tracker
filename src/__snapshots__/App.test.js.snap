// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue tracker loads even if user has no repos 1`] = `
<div
  className="loggedInWrapper"
>
  <h2>
    <svg
      alt=""
      aria-hidden="true"
      className="svg-inline--fa fa-code-branch fa-w-12 fa-1x "
      color="#CCC"
      data-icon="code-branch"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={
        Object {
          "marginRight": "1vw",
        }
      }
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M384 144c0-44.2-35.8-80-80-80s-80 35.8-80 80c0 36.4 24.3 67.1 57.5 76.8-.6 16.1-4.2 28.5-11 36.9-15.4 19.2-49.3 22.4-85.2 25.7-28.2 2.6-57.4 5.4-81.3 16.9v-144c32.5-10.2 56-40.5 56-76.3 0-44.2-35.8-80-80-80S0 35.8 0 80c0 35.8 23.5 66.1 56 76.3v199.3C23.5 365.9 0 396.2 0 432c0 44.2 35.8 80 80 80s80-35.8 80-80c0-34-21.2-63.1-51.2-74.6 3.1-5.2 7.8-9.8 14.9-13.4 16.2-8.2 40.4-10.4 66.1-12.8 42.2-3.9 90-8.4 118.2-43.4 14-17.4 21.1-39.8 21.6-67.9 31.6-10.8 54.4-40.7 54.4-75.9zM80 64c8.8 0 16 7.2 16 16s-7.2 16-16 16-16-7.2-16-16 7.2-16 16-16zm0 384c-8.8 0-16-7.2-16-16s7.2-16 16-16 16 7.2 16 16-7.2 16-16 16zm224-320c8.8 0 16 7.2 16 16s-7.2 16-16 16-16-7.2-16-16 7.2-16 16-16z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    GitHub Issue Tracker
    <span
      className="subtitle"
    >
       by justinh00k
    </span>
  </h2>
  <div
    className="wrapper"
  >
    <div
      className="column"
      id="repos"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <table>
        <tbody />
      </table>
    </div>
    <div
      className="column"
      id="issues"
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <table>
        <tbody>
          <tr>
            <td
              colSpan={4}
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`snapshot is unchanged since last 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <svg
      alt="Github Logo"
      aria-hidden="true"
      className="svg-inline--fa fa-github fa-w-16 fa-6x App-logo"
      color="#222222"
      data-icon="github"
      data-prefix="fab"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 496 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <h2>
      Please enter your Github 
      <a
        href="https://github.com/settings/tokens"
      >
        token
      </a>
      .
    </h2>
    <div
      className="loginWrapper"
    >
      <input
        className="loginInput"
        id="token"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Token"
        size={45}
        style={
          Object {
            "borderColor": "#999",
          }
        }
        value="dc1705a477c2f7b1c58b455c407debf0c015e377"
      />
      <svg
        alt="Github Login"
        aria-hidden="true"
        className="svg-inline--fa fa-github-alt fa-w-15 fa-3x loginButton"
        color="#999999"
        data-icon="github-alt"
        data-prefix="fab"
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 480 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </header>
</div>
`;
